pkg/sql/catalog/seqexpr/sequence.go

```diff
@@ -187,3 +190,123 @@ func ReplaceSequenceNamesWithIDs(
 	newExpr, err := tree.SimpleVisit(defaultExpr, replaceFn)
 	return newExpr, err
 }
+
+// UpgradeSequenceReferenceInExpr upgrades all by-name reference in `expr` to by-ID.
+// The name to ID resolution logic is aided by injecting a set of sequence names
+// (`allUsedSeqNames`). This set is expected to contain names of all sequences used
+// in `expr`, so all we need to do is to match each by-name seq reference in `expr`
+// to one entry in `allUsedSeqNames`.
+func UpgradeSequenceReferenceInExpr(
+	expr *string, allUsedSeqNames map[descpb.ID]*tree.TableName,
+) (hasUpgraded bool, err error) {
```

Comment:
I'm wondering if it wouldn't be simpler to have this function take a reverse index of type `map[string]descpb.ID` instead of `allUsedSeqNames map[descpb.ID]*tree.TableName`. This would move all of the logic explained in the commentary of `findUniqueBestMatchingForTableName` to the generation of the reverse index. Lookups could then be done directly via the `seqIdentifier.SeqName`. Ambiguous strings could simply have no entry (or map to a sentinel value like 0).